File: 00074. Search a 2D Matrix/00074.c
bool bs(int *nums, int beginIndex, int endIndex, int target) {
    int l = beginIndex, r = endIndex, half;
    
    while (true) {
        half = (l + r) / 2;
        
        if (nums[half] == target) {
            return true;
        } else if (half == l) {
            return nums[l] == target || nums[r] == target;
        } else if (nums[half] > target) {
            r = half-1;
        } else if (nums[half] < target) {
            l = half+1;
        }
    }
    
    return false;
}

bool searchMatrix(int **matrix, int matrixSize, int *matrixColSize, int target) {
    int numsSize = matrixSize * *matrixColSize;
    int *nums = (int *) malloc(sizeof(int) * numsSize);
    
    for (int i = numsSize - 1, col = *matrixColSize; i >= 0; --i) {
        nums[i] = matrix[i / col][i % col];
    }
    
    bool res = bs(nums, 0, numsSize - 1, target);
    
    free(nums);
    return res;
}

File: 00074. Search a 2D Matrix/00074.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = matrix.length;
        int col = matrix[0].length;
        int begin = 0, end = row * col - 1, half, halfValue;
        
        while (begin <= end) {
            half = (begin + end) / 2;
            halfValue = matrix[half / col][half % col];
                
            if (target == halfValue) {
                return true;
            } else if (target > halfValue) {
                begin = half+1;
            } else {
                end = half-1;
            }
        }
        
        return false;
    }
}
